src/combobox/ComboBox.ts

```diff
@@ -0,0 +1,389 @@
+			}
+			else {
+				this.selectResult(results[this._selectedIndex]);
+			}
+		},
+
+		[keys.escape](this: ComboBox) {
+			this.updateSelectedIndex(Operation.reset);
+			this._open = false;
+			this.invalidate();
+		}
+	};
+
+	updateSelectedIndex(operation?: Operation, index?: number) {
+		const { results } = this.properties;
+
+		if (!results || results.length === 0 || !this._open) {
+			return;
+		}
+
+		if (index !== undefined) {
+			this._selectedIndex = index;
+		}
+		else if (operation !== undefined) {
+			if (operation === Operation.increase || operation === Operation.decrease) {
+				const increment = operation === Operation.increase ? 1 : -1;
+				const total = results.length;
+				const current = this._selectedIndex;
+				const base = operation === Operation.increase ? -1 : 0;
+
+				this._selectedIndex = ((current === undefined ? base : current) + increment + total) % total;
+			}
+			else {
+				this._selectedIndex = undefined;
+			}
+			this._direction = operation;
+		}
+	}
+
+	createRegistry() {
+		const {
+			customResultItem = ResultItem,
+			customResultMenu = ResultMenu
+		} = this.properties;
+
+		const registry = new FactoryRegistry();
+		registry.define('result-item', customResultItem);
+		registry.define('result-menu', customResultMenu);
+		return registry;
+	}
+
+	@onPropertiesChanged
+	onPropertiesChanged(evt: PropertiesChangeEvent<this, ComboBoxProperties>) {
+		if (!this.registry || includes(evt.changedPropertyKeys, 'customResultItem') || includes(evt.changedPropertyKeys, 'customResultMenu')) {
+			this.registry = this.createRegistry();
+		}
+	}
+
+	afterCreate(element: HTMLElement) {
+		this._inputElement = <HTMLInputElement> element.querySelector('input');
+		// Maintain focused state
+		this._inputElement && this._inputElement[this._focused ? 'focus' : 'blur']();
+	}
+
+	afterUpdate(element: HTMLElement) {
+		// Maintain focused state
+		this._inputElement && this._inputElement[this._focused ? 'focus' : 'blur']();
+		// Make sure highlighted result is visible after arrow up / arrow down
+		const selectedResult = element.querySelector('[data-selected="true"]');
+		selectedResult && this.scrollIntoView(<HTMLElement> selectedResult);
+	}
+
+	scrollIntoView(element: HTMLElement) {
+		const menu = <HTMLElement> element.parentElement;
+		// Scroll menu up so top of highlighted result aligns with top of menu container
+		if (element.offsetTop - menu.scrollTop < 0) {
+			menu.scrollTop = element.offsetTop;
+		}
+		// Scroll menu down so bottom of highlighted result aligns with bottom of menu container
+		else if ((element.offsetTop - menu.scrollTop + element.offsetHeight) > menu.clientHeight) {
+			menu.scrollTop = element.offsetTop - menu.clientHeight + element.offsetHeight;
+		}
+	}
+
+	onInput(event: Event) {
+		const {
+			onRequestResults,
+			onChange
+		} = this.properties;
+
+		this.updateSelectedIndex(Operation.reset);
+		this._open = true;
+		onChange && onChange((<HTMLInputElement> event.target).value);
```

Comment:
Would be good if we created a type this, I think one used to exist in `widget-core`